### PR TITLE
Map rendering fix

### PR DESCRIPTION
**Pull request by baughj**
Fixes rendering error for non-isometric maps. Also adds `MapImageCache` as a parameter to `RenderMap` to allow a shared cache to be used (I intend to use it for rendering multiple maps at a time, because doing the individual tile renderings over and over is suboptimal). I also made the padding configurable.

### CODE REVIEW

File: DALib/Drawing/Graphics.cs [hunk 1]
```diff
@@ -193,43 +193,50 @@ public static SKImage RenderMap(MapFile map, DataArchive seoDat, DataArchive iaD
...
         //create lookups so we only render each tile piece once
         using var bgCache = new SKImageCache<int>();
         using var lfgCache = new SKImageCache<int>();
         using var rfgCache = new SKImageCache<int>();
 
-        //calculate width and height based on isometric view
-        var width = map.Width * CONSTANTS.TILE_WIDTH;
-        var height = map.Height * (CONSTANTS.TILE_HEIGHT + 1) + FOREGROUND_PADDING;
+        //calculate width and height
+        var width = (map.Width + map.Height + 1) * CONSTANTS.HALF_TILE_WIDTH;
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs:242]: collateral correction for below correction

```suggestion
        var width = (map.Width + map.Height) * CONSTANTS.HALF_TILE_WIDTH;
```

File: DALib/Drawing/Graphics.cs [hunk 2]
```diff
@@ -193,43 +193,50 @@ public static SKImage RenderMap(MapFile map, DataArchive seoDat, DataArchive iaD
...
+        var height = (map.Width + map.Height) * CONSTANTS.HALF_TILE_HEIGHT + foregroundPadding;
         using var bitmap = new SKBitmap(width, height);
         using var canvas = new SKCanvas(bitmap);
 
         //the first tile drawn is the center tile at the top (0, 0)
-        var bgInitialDrawX = width / 2 - CONSTANTS.HALF_TILE_WIDTH;
-        var bgInitialDrawY = FOREGROUND_PADDING;
+        var bgInitialDrawX = map.Height * CONSTANTS.HALF_TILE_WIDTH;
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs]: the first tile should be drawn at an offset of 0

```suggestion
        var bgInitialDrawX = (map.Height - 1) * CONSTANTS.HALF_TILE_WIDTH;
```

File: DALib/Drawing/Graphics.cs [hunk 3]
```diff
@@ -193,43 +193,50 @@ public static SKImage RenderMap(MapFile map, DataArchive seoDat, DataArchive iaD
...
         MapFile map,
         Tileset tiles,
         PaletteLookup bgPaletteLookup,
         PaletteLookup fgPaletteLookup,
-        DataArchive iaDat)
+        DataArchive iaDat, 
+        int foregroundPadding = 512,
+        MapImageCache? cache = null)
     {
-        const int FOREGROUND_PADDING = 512;
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs:235]: could do something like this...
```suggestion
//only dispose the cache if it was created in this method
var dispose = cache is null;
cache ??= new MapImageCache();

try
{
    ...
} finally 
{
   if (dispose)
       cache.Dispose();
}
```

File: DALib/Drawing/Graphics.cs [hunk 4]
```diff
@@ -241,7 +248,8 @@ public static SKImage RenderMap(
                 if (bgIndex > 0)
                     --bgIndex;
 
-                var bgImage = bgCache.GetOrCreate(
+                
+                var bgImage = (cache == null ? bgCache : cache.BackgroundCache).GetOrCreate(
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs]: if you use the try/finally from above, can skip the ternary

File: DALib/Drawing/Graphics.cs [hunk 5]
```diff
@@ -275,7 +283,7 @@ public static SKImage RenderMap(
                 var rfgIndex = tile.RightForeground;
 
                 //render left foreground
-                var lfgImage = lfgCache.GetOrCreate(
+                var lfgImage = (cache == null ? lfgCache : cache.LeftForegroundCache).GetOrCreate(
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs]: if you use the try/finally from above, can skip the ternary

File: DALib/Drawing/Graphics.cs [hunk 6]
```diff
@@ -294,7 +302,7 @@ public static SKImage RenderMap(
                     canvas.DrawImage(lfgImage, lfgDrawX, lfgDrawY);
 
                 //render right foreground
-                var rfgImage = rfgCache.GetOrCreate(
+                var rfgImage = (cache == null ? rfgCache : cache.RightForegroundCache).GetOrCreate(
```
Comment (Sichii) [inline DALib/Drawing/Graphics.cs]: if you use the try/finally from above, can skip the ternary

File: DALib/Utility/MapImageCache.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    /// <inheritdoc/>
+    public virtual void Dispose()
+    {
+        BackgroundCache.Dispose();
+        LeftForegroundCache.Dispose();
+        RightForegroundCache.Dispose();
+        GC.SuppressFinalize(this);
+    }
+}
```
Comment (Sichii) [inline DALib/Utility/MapImageCache.cs]:
```suggestion
public class MapImageCache: IDisposable
{
    /// <summary>
    ///     Initializes a new instance of the <see cref="MapImageCache"/> class.
    /// </summary>
    public MapImageCache()
    {
        BackgroundCache = new SKImageCache<int>();
        LeftForegroundCache = new SKImageCache<int>();
        RightForegroundCache = new SKImageCache<int>();
    }

    /// <summary>
    ///     Initializes a new instance of the <see cref="MapImageCache"/> class.
    /// </summary>
    public MapImageCache(SKImageCache<int> bgCache, SKImageCache<int> lfgCache, SKImageCache<int> rfgCache)
    {
        BackgroundCache = bgCache;
        LeftForegroundCache = lfgCache;
        RightForegroundCache = rfgCache;
    }
    
    /// <summary>
    /// The background cache
    /// </summary>
    public SKImageCache<int> BackgroundCache { get; }
    /// <summary>
    /// The left foreground cache
    /// </summary>
    public SKImageCache<int> LeftForegroundCache { get; }
    /// <summary>
    /// The right foreground cache
    /// </summary>
    public SKImageCache<int> RightForegroundCache { get; }

    /// <inheritdoc/>
    public virtual void Dispose()
    {
        BackgroundCache.Dispose();
        LeftForegroundCache.Dispose();
        RightForegroundCache.Dispose();
        GC.SuppressFinalize(this);
    }
}
```

**Review by Sichii — COMMENTED** [review]
just some thoughts, works either way

File: DALib/Drawing/Graphics.cs (hunk 1 above)
Comment (baughj) [reply DALib/Drawing/Graphics.cs:242]: fixed

File: DALib/Drawing/Graphics.cs (hunk 2 above)
Comment (baughj) [reply DALib/Drawing/Graphics.cs]: fixed

**Comment by baughj** [comment]
@Sichii thanks, all good suggestions! merging

**Pull request merged**